presto-main/src/main/java/com/facebook/presto/sql/MaterializedViewUtils.java

```diff
@@ -175,4 +195,83 @@ public static TupleDomain<String> getDomainFromFilter(Session session, DomainTra
 
         return predicateFromBaseQuery.getTupleDomain();
     }
+
+    /**
+     * Rewrites the provided argument if it is non-associative (e.g. it cannot be handled by simply re-applying function to derived columns)
+     */
+    public static Optional<Expression> rewriteNonAssociativeFunction(FunctionCall functionCall, Map<Expression, Identifier> baseToViewColumnMap)
+    {
+        if (functionCall.isDistinct() || functionCall.getWindow().isPresent() || functionCall.getOrderBy().isPresent()
+                || functionCall.isIgnoreNulls() || functionCall.getFilter().isPresent()) {
+            return Optional.empty();
+        }
+
+        QualifiedName functionName = functionCall.getName();
+        List<Expression> expressions = functionCall.getArguments();
+
+        if (!NON_ASSOCIATIVE_REWRITE_FUNCTIONS.containsKey(functionName) || expressions.size() > 1 || !(expressions.get(0) instanceof Identifier)) {
+            return Optional.empty();
```

Comment:
Instead of returning empty, error message or exception can provide more context to the user?